python/mlx/nn/layers/flatten.py

```diff
@@ -0,0 +1,61 @@
+import mlx.core as mx
+from mlx.nn.layers.base import Module
+
+
+def flatten(a: mx.array, start_dim: int = 0, end_dim: int = -1) -> mx.array:
+    r"""
+    Flattens a contiguous range of dims into a tensor.
+
+    See :class:`~nn.Flatten` for details.
+
+    Args:
+        input: input tensor.
+        start_dim: first dim to flatten (default = 0).
+        end_dim: last dim to flatten (default = -1).
+    """
+
+    return mx.flatten(a, start_dim, end_dim)
+
+
+class Flatten(Module):
+    r"""
+    Flattens input by reshaping it into a one-dimensional array. For use with :class:`~nn.Flatten`.
```

Comment:
I'm curious about your use case for this Module? Why not just do the reshape / flatten in the `__call__` function of your container Module? 

We are trying to avoid Module bloat (e.g. implementing a Module wrapper for every op in mlx.core) so to the extent that it's easy to just use the ops directly we prefer to go that route and keep the top-level `nn` clutter-free.

Reply:
The use of nn.Flatten or similars like reshape in mlx is a matter of convenience and clarity in code rather than a strict necessity. The choice between using a specific module for flattening or doing the operation directly in the forward function depends on the specific requirements and design preferences.

for example we can use the flatten Module with the Sequential Module as follows:
```python
import mlx.core as mx
import mlx.nn as nn

sequential_net = nn.Sequential(
    nn.Conv2d(1, 32, kernel_size=3, stride=1, padding=1),
    nn.ReLU(),
    nn.Flatten(),
    nn.Linear(32 * 28 * 28, 10)  # Assuming input image size is 28x28
)

input_tensor = mx.random.normal((16, 1, 28, 28))  # Batch size of 1, 1 channel, 28x28 image

# Forward pass
output = sequential_net(input_tensor)
```

Reply:
Yes, I certainly see the point. Nevertheless, I think it's a slippery slope to go this route as we have to start mirroring all of our ops with modules. I would prefer if you remove this for now and just land the `mx.flatten` op while we think about the right approach here.

For example, maybe something like a lambda module is a lot cleaner `nn.Lambda(fun)` which calls `fun` on it forward.

Reply:
Also the `nn.Flatten` module is designed to preserve the batch dimension while reshaping the input tensor into a one-dimensional format. It achieves this by collapsing all dimensions, except for the batch dimension. This eliminates the need to use `mx.flatten(a, start_dim=1)` for the same flattening operation. WDYT?

Reply:
Yea that’s a slight convenience but let’s hold off on adding it for now. Do you mind splitting it out of this PR? We will revisit the module decision after some discussion.

Reply:
Ahah, uure, I'll take care of removing the agreed-upon stuff and reach out to you.